### PR TITLE
(Fixes #12182) Replaced urlencoded OAuth2 creds with Basic header

### LINKED ISSUE #12182: [OAuth2.0] Token request should use Basic auth instead of urlencoded credentials
**Describe the bug**
The OAuth2 token request should use Basic auth instead of urlencoded credentials.  [RFC 6749 section 2.3.1](https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1) (The OAuth 2.0 Authorization Framework)  states:
```
Including the client credentials in the request-body using the two
   parameters is NOT RECOMMENDED and SHOULD be limited to clients unable
   to directly utilize the HTTP Basic authentication scheme
```

**Expected behavior**
The OAuth2 token request should use the "Authorization: Basic ..." header for Oauth2 `client_id` and `client_secret` credential exchange.

**Additional context**
Code where client creds are being put into the body of the token request:  https://github.com/apache/pulsar/blob/v2.8.1/pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/oauth2/protocol/TokenClient.java#L76-L77

### PR DESCRIPTION
**Pull request by danielorf**
Fixes #12182

### Motivation
Token request should use Basic auth instead of urlencoded credentials

### Modifications

Replaced the URLEncoded OAuth2 creds with the `Authorization Basic ...` variant in accordance with [RFC 6749 section 2.3.1](https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1)  recommendation.

Note that this could affect users of OAuth2 where their token provider accepts URLEncoded credentials but not `Authorization Basic ...` header - which would be against the recommendation of [RFC 6749 section 2.3.1](https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1).

### Verifying this change

- [x] Make sure that the change passes the CI checks.
- Updated TokenClientTest to reflect the cred request change

### Documentation

Check the box below and label this PR (if you have committer privilege).

Need to update docs? 

- [X] no-need-doc:   **Simply correcting the request type when requesting a token from OAuth2 provider**

### CODE REVIEW

**Comment by danielorf** [comment]
Can someone with access to CI restart the pipeline?  It appears that one of the jobs was canceled (but the rest passed).

**Comment by danielorf** [comment]
Good to know, thanks.  Is this documented anywhere?

**Comment by danielorf** [comment]
I see what looks like an unusual error in the `CI - Unit - Brokers - Broker Group 1` / `run unit test 'BROKER_GROUP_1'` stage:

```
[INFO] Running org.apache.pulsar.utils.StatsOutputStreamTest
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.008 s - in org.apache.pulsar.utils.StatsOutputStreamTest
Warning:  Tests run: 29, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 1,489.704 s - in org.apache.pulsar.broker.service.ReplicatorTest
[INFO] Tests run: 70, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 482.895 s - in org.apache.pulsar.broker.admin.TopicPoliciesTest
Error: The operation was canceled.
```

I'm going to re-run again but may need some help debugging the test environment.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/oauth2/protocol/TokenClient.java [hunk 1]
```diff
@@ -96,10 +98,9 @@ String buildClientCredentialsBody(Map<String, String> bodyMap) {
      */
     public TokenResult exchangeClientCredentials(ClientCredentialsExchangeRequest req)
             throws TokenExchangeException, IOException {
+        String credPayload = req.getClientId() + ":" + req.getClientSecret();
         Map<String, String> bodyMap = new TreeMap<>();
         bodyMap.put("grant_type", "client_credentials");
-        bodyMap.put("client_id", req.getClientId());
-        bodyMap.put("client_secret", req.getClientSecret());
```
Comment (lmccay) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/oauth2/protocol/TokenClient.java:102]: Is there a backward compatibility issue here that could be avoided with client side config, sys property or something?

**Comment by tisonkun** [comment]
Closed as stale and conflict. Please rebase and resubmit the patch if it's still relevant.

**Pull request closed without merging**